### ISSUE TITLE
About future open-source support for continued MASK pretraining in NTv3

### ISSUE THREAD

**Issue by Liujh0223** [instadeepai/nucleotide-transformer#131]
Hi NTv3 team,

Thank you for open-sourcing NTv3 — it’s a genuinely impressive and exciting genomic language model. I noticed you’ve already released workflows for fine-tuning using BigWig tracks and annotation files, which is incredibly helpful.

I was wondering whether you might also plan to provide open-source support for continued pretraining (i.e., further training using the masked self-supervised/MASK objective) in the future. If so, any pointers to an expected roadmap or recommended best practices would be greatly appreciated.

Thanks again for your work and for making these resources available.

Best regards,

**Comment by ranzenTom** [comment]
Hi,

Thanks for the kind words about NTv3!

Regarding continued pretraining: we won't be releasing our internal data pipeline as it's tied to our infrastructure. However, since NTv3 is implemented as AutoModelForMaskedLM, you can use the standard HuggingFace MLM training utilities directly.

Here's a minimal example to get started with continued pretraining:

```python
from transformers import AutoTokenizer, AutoModelForMaskedLM, DataCollatorForLanguageModeling, Trainer, TrainingArguments
from datasets import Dataset

# Load model and tokenizer
repo = "InstaDeepAI/NTv3_100M_pre"
tokenizer = AutoTokenizer.from_pretrained(repo, trust_remote_code=True)
model = AutoModelForMaskedLM.from_pretrained(repo, trust_remote_code=True)

# Prepare your sequences (replace with your actual data)
sequences = ["ATCGATCGATCG", "GCTAGCTAGCTA"]  # your DNA sequences
dataset = Dataset.from_dict({"text": sequences})

def tokenize_function(examples):
    return tokenizer(
        examples["text"],
        add_special_tokens=False,
        padding="max_length",
        truncation=True,
        max_length=1024,  # must be multiple of 128
        return_tensors="pt"
    )

tokenized_dataset = dataset.map(tokenize_function, batched=True, remove_columns=["text"])

# MLM data collator handles masking automatically
data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=True,
    mlm_probability=0.15
)

# Training
training_args = TrainingArguments(
    output_dir="./ntv3-continued-pretraining",
    per_device_train_batch_size=8,
    learning_rate=1e-5,
    num_train_epochs=3,
    bf16=True,  # if using A100/H100
    report_to="none" #Change to your configured logging service 
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset,
    data_collator=data_collator,
)

trainer.train()
```

Key points:

- Input sequence length must be a multiple of 128
- We recommend bf16 for efficient training on A100/H100 GPUs
- The DataCollatorForLanguageModeling handles dynamic masking

Best,
Thomas

**Comment by Liujh0223** [comment]
Understood. Thank you very much for Thomas's explanation. Best wishes.
